src/gui/transferlistmodel.cpp

```diff
@@ -663,17 +666,30 @@ BitTorrent::Torrent *TransferListModel::torrentHandle(const QModelIndex &index)
 
 void TransferListModel::handleTorrentAboutToBeRemoved(BitTorrent::Torrent *const torrent)
 {
-    const int row = m_torrentMap.value(torrent, -1);
+    const qsizetype row = m_torrentMap.value(torrent, -1);
     Q_ASSERT(row >= 0);
 
     beginRemoveRows({}, row, row);
     m_torrentList.removeAt(row);
     m_torrentMap.remove(torrent);
-    for (int &value : m_torrentMap)
+
+    // pick an efficient method
+    const qsizetype shiftAffectedCost = (m_torrentList.size() - row) * 2;  // 2 lookups
```

Comment:
It doesn't still look as more or less correct estimation.
Iterating over `m_torrentList` and over `m_torrentMap` is more or less comparable so you can assign the same cost to the iteration step. But while iterate over `m_torrentMap` you don't do some expensive work on each step (only test Boolean variable). While iterating over `m_torrentList` you perform lookup value from `m_torrentMap` that at least does hashing of key and searching of this key in index (in search tree, I suppose, so it is something like log2(n)/2 steps in average). So assigning the same cost to lookup in `m_torrentMap` and step of iteration is incorrect, IMO.

Reply:
>It doesn't still look as more or less correct estimation.
Iterating over m_torrentList and over m_torrentMap is more or less comparable so you can assign the same cost to the iteration step. But while iterate over m_torrentMap you don't do some expensive work on each step (only test Boolean variable). While iterating over m_torrentList you perform lookup value from m_torrentMap that at least does hashing of key and searching of this key in index (in search tree, I suppose, so it is something like log2(n)/2 steps in average). So assigning the same cost to lookup in m_torrentMap and step of iteration is incorrect, IMO.

OK, but please suggest something concrete and can be computed cheaply.
I don't really care about the exact cost as long as it will pick the less-work method in the case of `row` near to the end of total rows.

Reply:
> OK, but please suggest something concrete and can be computed cheaply.
> I don't really care about the exact cost as long as it will pick the less-work method in the case of `row` is near to the end of total rows.

```c++
struct Item
{
    Torrent *torrent;
    int row;
};
QHash<Torrent *, std::shared_ptr<Item>> m_utensByTorrent; // instead of m_torrentMap;
QList<std::shared_ptr<Item>> m_items; // instead of m_torrentList;
```
Then you just need to iterate over `m_items` and change `Item::row` without extra lookup.

Reply:
Sorry I can't comprehend what you are writing.
I'll give an usage scenario: an user has 10000 torrents and now he removes the torrent at row 9999. ALL of your suggestions seems to iterate over (and doing comparison along) 10000 elements, how is that more efficient than just removing row 9999 and subtract 1 from row 10000?

Reply:
> Sorry I can't comprehend what you are writing.
ALL of your suggestions seems to iterate over (and doing comparison along) 10000 elements

Why?
> Then you just need to iterate over `m_items` and change `Item::row` without extra lookup.

I mean iterate from given row to the end.
```c++
for (auto it = m_items.begin() + row; it != m_items.end(); ++it)
    --*it->row;
```

Reply:
>It doesn't still look as more or less correct estimation.

I'll propose change it to:
```c++
const qsizetype shiftAffectedCost = (m_torrentList.size() - row) * 3;  // 2 lookups + 1 hashing
```
Those are just simple operation counts.
If you still doesn't think it is good enough then I'll drop this commit, as I want to back port the 'UI hang' commit to v4.5 and I don't want to wait forever.

Reply:
```
for (auto it = m_items.begin() + row; it != m_items.end(); ++it)
```
How many elements does it loop given that scenario?

Reply:
> I mean iterate from given row to the end.
> 
> ```c++
> for (auto it = m_items.begin() + row; it != m_items.end(); ++it)
>     --*it->row;
> ```

This is exactly the same as my code:
```c++
        for (qsizetype i = row; i < m_torrentList.size(); ++i)
            --m_torrentMap[m_torrentList[i]];
```
Aren't you aware you are dereferencing a shared pointer?

Reply:
> Aren't you aware you are dereferencing a shared pointer?

No. I was away from the computer to check it out. In fact, it should be like this:
```c++
for (auto it = m_items.begin() + row; it != m_items.end(); ++it)
    --(*it)->row;    
```

Reply:
> I mean it seems to contradicts your claim: `Wasn't the previous way more effective? Iterating the QHash should be much less overhead compared to repeatedly lookup value by key.`

Sorry, I don't understand what contradicts what.

Your code performs lookup value by key (`m_torrentMap[some_key]`) whereas my not but just dereference the pointer and access the field of struct (`(*it)->row`). If you haven't gotten too carried away with high-level languages (JavaScript, Python etc.) lately so that you've forgotten what's going on "under the hood", then you should understand that these are operations of different complexity.

Reply:
#17998 implements my suggestion.

Reply:
>https://github.com/qbittorrent/qBittorrent/pull/17998 implements my suggestion.

Quoting myself:
>Besides, you introduced additional lookup latency when looking up from row number or torrent pointer. IMO additional latency is only tolerable when adding/removing torrents.

There you made look up slower and uses more memory with the benefit of only removing faster. IMO it is not a great tradeoff.

Reply:
Well, I still did not fail to do some testing:

```c++
#include <QDebug>
#include <QHash>
#include <QList>
#include <QDateTime>

int main()
{
    QList<int> list;
    QHash<int, int> map;

    for (int i = 0; i < 100000; ++i)
    {
        int val = 100000 - i;
        list.append(val);
        map[val] = i;
    }

    int lim = 50000;
    for (int ii = 0; ii < 10; ++ii)
    {
        qint64 t1 = QDateTime::currentMSecsSinceEpoch();
        for (int &i : map)
        {
            if (i > lim)
                --i;
        }
        qDebug() << "Time1:" << QDateTime::currentMSecsSinceEpoch() - t1;

        qint64 t2 = QDateTime::currentMSecsSinceEpoch();
        for (int i = lim; i < list.size(); ++i)
            --map[list[i]];
        // The following will increase it
//        for (auto it = list.cbegin() + lim; it != list.cend(); ++it)
//            --map[*it];
        qDebug() << "Time2:" << QDateTime::currentMSecsSinceEpoch() - t2;
        qDebug() << "";
    }
}
```

The results say that the search for a value in QHash is quite optimized (more than I thought). Honestly, I used integers as keys, but we have pointers, so it shouldn't be much different, as, say, if we had strings.
Your algorithm starts to outperform when processing about 40-35% of the entire list. Interestingly, if you use iterators (as in commented code), it accelerates significantly, so that it starts to outperform near 50% of list. I suggest you take independent measurements using my test.

It turns out that you can adjust the solution you proposed above and use it.